specs/core/1_beacon-chain-misc.md

```diff
@@ -0,0 +1,218 @@
+        - [committee_to_compact_committee](#committee_to_compact_committee)
+        - [get_previous_power_of_2](#get_previous_power_of_2)
+        - [verify_merkle_proof](#verify_merkle_proof)
+        - [concat_generalized_indices](#concat_generalized_indices)
+        - [compute_historical_state_generalized_index](#compute_historical_state_generalized_index)
+        - [get_generalized_index_of_crosslink_header](#get_generalized_index_of_crosslink_header)
+        - [process_shard_receipt](#process_shard_receipt)
+    - [Changes](#changes)
+        - [Persistent committees](#persistent-committees)
+        - [Shard receipt processing](#shard-receipt-processing)
+
+<!-- /TOC -->
+
+## Classes
+
+#### `CompactCommittee`
+
+```python
+class CompactCommittee(Container):
+    pubkeys: List[BLSPubkey, MAX_VALIDATORS_PER_COMMITTEE]
+    compact_validators: List[uint64, MAX_VALIDATORS_PER_COMMITTEE]
+```
+
+#### `ShardReceiptProof`
+
+```python
+class ShardReceiptProof(Container):
+    shard: Shard
+    proof: List[Hash, PLACEHOLDER]
+    receipt: List[ShardReceiptDelta, PLACEHOLDER]
+```
+
+## Helpers
+
+#### `pack_compact_validator`
+
+```python
+def pack_compact_validator(index: uint64, slashed: bool, balance_in_increments: uint64) -> uint64:
+    """
+    Creates a compact validator object representing index, slashed status, and compressed balance.
+    Takes as input balance-in-increments (// EFFECTIVE_BALANCE_INCREMENT) to preserve symmetry with
+    the unpacking function.
+    """
+    return (index << 16) + (slashed << 15) + balance_in_increments
+```
+
+#### `unpack_compact_validator`
+
+```python
+def unpack_compact_validator(compact_validator: uint64) -> Tuple[uint64, bool, uint64]:
+    """
+    Returns validator index, slashed, balance // EFFECTIVE_BALANCE_INCREMENT
+    """
+    return compact_validator >> 16, (compact_validator >> 15) % 2, compact_validator & (2**15 - 1)
+```
+
+#### `committee_to_compact_committee`
+
+```python
+def committee_to_compact_committee(state: BeaconState, committee: Sequence[ValidatorIndex]) -> CompactCommittee:
+    """
+    Given a state and a list of validator indices, outputs the CompactCommittee representing them.
+    """
+    validators = [state.validators[i] for i in committee]
+    compact_validators = [
+        pack_compact_validator(i, v.slashed, v.effective_balance // EFFECTIVE_BALANCE_INCREMENT)
+        for i, v in zip(committee, validators)
+    ]
+    pubkeys = [v.pubkey for v in validators]
+    return CompactCommittee(pubkeys=pubkeys, compact_validators=compact_validators)
+```
+
+#### `get_previous_power_of_2`
+
+```python
+def get_previous_power_of_2(x: int) -> int:
+    return x if x <= 2 else 2 * get_previous_power_of_2(x // 2)
+```
+
+#### `verify_merkle_proof`
+
+```python
+def verify_merkle_proof(leaf: Hash, proof: Sequence[Hash], index: GeneralizedIndex, root: Hash) -> bool:
+    assert len(proof) == get_generalized_index_length(index)
```

Comment:
Undefined `get_generalized_index_length`

Reply:
Generalized index-related methods are defined here: https://github.com/ethereum/eth2.0-specs/pull/1186/files#diff-fb626d711756bc06ef07ecd51b009215R151

So #1186 is a dependency of this PR.